### ISSUE TITLE
Any advice on ODCI error

### ISSUE THREAD

**Issue by SEPTEMBER998767** [pljson/pljson#112]
Hi, Firstly thanks for the lib.
Done successfully installed on 12c (which is development server), (target deployment will be 11g).
After installed, testing and executing pljson table facing following error. 

ORA-29913: error in executing ODCITABLEDESCRIBE callout
Any advice.

![odcierr](https://user-images.githubusercontent.com/24283280/35324142-d118fb6e-012a-11e8-950f-463ba3a4f8ce.jpg)

**Comment by dsnz** [comment]
Hi,

It works fine for me (I retested a few minutes ago just in case using Oracle 12.2.0.1 and within both sqlplus and SQL Developer 17.2.0.188).
So we have to find what is different in your installation.

If you can supply the following information, it would be very helpful.

1. your exact Oracle version (all 4-5 numbers)
2. try running @examples\ex_pljson_table.sql from sqlplus
3. do step 2. in the Oracle 11 installation
(just in case.. even though I think this should go fine, the error you sent is most probable on 12 only)
4. the result of

select object_name, object_type, status
from user_objects
order by status, object_name, object_type

**Comment by dsnz** [comment]
Hi,

You are on 12.1 while I'm on 12.2, but I don't think this is your problem.
Other ideas

1. you did not say clearly if you tried running

```@examples/ex_pljson_table.sql```

within sqlplus and logged in as the user which is the owner of the PLJSON installation

please verify exact setup of test and reply with your result

2. have you made any custom changes to the pljson installation ? (some change to code or synonyms)
3. when you run the select, do you have full access to tables and packages used ?
(and I mean full access with direct grant and not via a role !
in fact this is the most common reason for the message you get: no full access to underlying tables and packages)
4. besides the screenshot error, do you get other error information ? (line number ?)

**Comment by SEPTEMBER998767** [comment]
Dear D,
Am suspecting access grant as well, but since i'm not known what to ask to DAB, DB team will shoot me back to check myself :dancing_men: 

1: Pls refer below
2: No, I didn't change any
3: I think login user has full access, 
    Role grant, so far i check is "NO", its by user
4: line no 1 though :1st_place_medal: 

![2018-01-26_9-05-10](https://user-images.githubusercontent.com/24283280/35420929-352d03ce-027b-11e8-8680-34469da81d17.jpg)
![2018-01-26_9-05-26](https://user-images.githubusercontent.com/24283280/35420935-375e0454-027b-11e8-9a00-4b489d16fb09.jpg)

Thank you.

**Comment by dsnz** [comment]
Hello,

As you understand it's very difficult to support you with this issue which has to do with your database, especially since I cannot have direct access. However solving it may help other users in a similar situation in the future and I would like to give it a try again. But you must get the help of a DBA and do some steps together as I describe below. I guess it should be easy since we are talking about a development database.

Do these steps in order and in short time.

```
(first by a dba)
1. grant alter session to <owner of PLJSON installation>;
2. alter system flush shared_pool; (this step cannot be skipped !)

(then as owner of PLJSON installation)
3. make a fresh new login
4. alter session set tracefile_identifier = 'pljson';
5. alter session set events '10046 trace name context forever, level 12';
6. run the sql statement you posted here
7. logoff
```

then your DBA will look at the user dump folder and find 2 trace files named

```
<db name>_ora_<nnnn>_pljson.trc and
<db name>_ora_<nnnn>_pljson.trm
```

(they can be named differently in a very customized db setup, but the file name will always end in 'pljson' and the files should be recent files at the same time of your test run)

please zip these files and post here
(do not edit these files as I will use a trace analyzer tool to view the contents;
the files don't contain sensitive content of your installation but if you are still anxious just email the zip to me)

Thank you and I hope that with these trace files we can understand what's happening and resolve this issue.

**Comment by serkangh** [comment]
Hello,
I have the same issue.
The owner of this issue don't reply to your last request.
But I collected the trace files as zip. And attached to this issue.

We have one development and one production database all 11.2.0.4.0.
We tested in development and works well.
Then try to install to production with success. But the queries are giving the same error.
ORA-29913 ODCITABLEDESRIBE

Can you help with this trace files.

Note: We create a fresh user to hold this packages;
-- Create the user 
create user WEB_JSON
  default tablespace BASKENT_DATA
  temporary tablespace TEMP
  profile DEFAULT
  password expire;
-- Grant/Revoke role privileges 
grant connect to WEB_JSON;
grant resource to WEB_JSON;
-- Grant/Revoke system privileges 
grant alter session to WEB_JSON;
grant create any synonym to WEB_JSON;
grant unlimited tablespace to WEB_JSON;

[Archive.zip](https://github.com/pljson/pljson/files/2025226/Archive.zip)

**Comment by serkangh** [comment]
We are queuing this dummy son with sql.

select * from table(pljson_table.json_table(
  '[
      {
      "ID": 780,
      "Ad": "1.Exam",
      "Tarih": "05.032017"
   },
      {
      "ID": 805,
      "Ad": "2.Exam",
      "Tarih": "09.072017"
   }
]',
  pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
  pljson_varray('ID', 'Ad', 'Tarih'),
  table_mode => 'nested'
));

**Comment by dsnz** [comment]
Hi and thank you,
I will have a look at it until the next day.
[sent from mobile]

**Comment by dsnz** [comment]
Hello,

1. you said that in 'development database' you have no problem
but in 'production database' you get the ORA-29913 error ??

2. please provide the output of
```
select object_name, object_type, status
from user_objects
where object_name like 'PLJ%'
order by object_name, object_type
```
and
```
select * from all_synonyms
where synonym_name like 'PLJ%'
```

**Comment by serkangh** [comment]
[SQLfirst.xlsx](https://github.com/pljson/pljson/files/2028664/SQLfirst.xlsx)
[SQLsecond.xlsx](https://github.com/pljson/pljson/files/2028665/SQLsecond.xlsx)

The outputs are in these excel files.

And my answer for your first question is YES.
Development server is running well.
Production server is giving this error.

Thanks.

P.S. : This issue is seen as closed. Can you reopen it again?

**Comment by dsnz** [comment]
Please flush the shared pool and try again the same example you posted which I repeat correctly (try this immediately)
```
select * from table(pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
pljson_varray('ID', 'Ad', 'Tarih'),
table_mode => 'nested'
));
```

**Comment by serkangh** [comment]
I flushed but the same error accord again.

As DBA
Connected to Oracle Database 11g Enterprise Edition Release 11.2.0.4.0 
Connected as opuser@RACDB1

SQL> alter system flush shared_pool;
System altered

As Web_Json user
Connected to Oracle Database 11g Enterprise Edition Release 11.2.0.4.0 
Connected as web_json@racdb1

SQL> 
SQL> select * from table(pljson_table.json_table(
  2  '[
  3  {
  4  "ID": 780,
  5  "Ad": "1.Exam",
  6  "Tarih": "05.032017"
  7  },
  8  {
  9  "ID": 805,
 10  "Ad": "2.Exam",
 11  "Tarih": "09.072017"
 12  }
 13  ]',
 14  pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
 15  pljson_varray('ID', 'Ad', 'Tarih'),
 16  table_mode => 'nested'
 17  ));
select * from table(pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
pljson_varray('ID', 'Ad', 'Tarih'),
table_mode => 'nested'
))
ORA-29913: ODCITABLEDESCRIBE başvurusunu yürütme hatası

**Comment by dsnz** [comment]
Hello,

After extensive checking of the trace file you sent, I am very puzzled. It appears to be a not consistent trace sequence of events and it also seems as if you executed an sql statement with bind variables (unlike the statement you posted) and most probably a statement like
```
select * from table(pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
null,
null,
null
));
```

or something similar (since I can't see the bind variables !) and which of course will produce ORA-29913 as it is invalid (because of the null path parameter)

so, the only way I can see to help you and troubleshoot this issue is to give me remote access to your development database (WEB_JSON schema only of course) and do my own tests and troubleshooting

**Comment by serkangh** [comment]
Ok.
I can give you access to our database.
Now the time is 22:20 with GMT+3
I send a request to info@sarmas.com from Skype app.

**Comment by dsnz** [comment]
Hi,
Just send me an email with details of your preferred method of access
e.g direct network access to port 1521 (this seems easier to start with and allows access only to WEB_JSON without messing with OS and user account permissions), or teamviewer or openvpn or ...
or probably the standard remote access method your site uses (can skype be used ?? I don't know)
so that is workable for me and safe for your site too
unfortunately due to my being busy most time I cannot provide you ahead the time I will connect (or I may connect many times trying many things at very strange times...) so you must allow me a way to do it anytime within the next two days (until Friday-Saturday)
we'd better arrange this via email than discussing here so contact me

**Comment by serkangh** [comment]
Hi,
I have 3 development and 1 prod database.
All of them 11.2.0.4.
All the development databases can run this sql query successful.
But prod can not.
I dropped web_json user and try to recreate schema and install.sql without errors.
And the error continues.

I can give you all time access to development servers but they are all running well.
I can not give the same all time access to prod database. Because of my company system admin permissions.

If you e-mail me that you can work on the issue on our prod database. ( I can check the e-mail frequently)
I can open a team viewer over my laptop to prod database.
Or you can leave a mini call. When I saw you I can open laptop in a 5 minutes. 😊

NOT: The diff with prod and dev systems is prod system is a RAC server. The others are single instance databases.

Thanks for your great help top issue and open source community.

Serkan AKCIN
Mobile +90 533 600 50 84
________________________________
Gönderen: E.I.Sarmas <notifications@github.com>
Gönderildi: 23 Mayıs 2018 Çarşamba 22:53
Kime: pljson/pljson
Bilgi: serkangh; Comment
Konu: Re: [pljson/pljson] Any advice on ODCI error (#112)


Hi,
Just send me an email with details of your preferred method of access
e.g direct network access to port 1521 (this seems easier to start with and allows access only to WEB_JSON without messing with OS and user account permissions), or teamviewer or openvpn or ...
or probably the standard remote access method your site uses (can skype be used ?? I don't know)
so that is workable for me and safe for your site too
unfortunately due to my being busy most time I cannot provide you ahead the time I will connect (or I may connect many times trying many things at very strange times...) so you must allow me a way to do it anytime within the next two days (until Friday-Saturday)
we'd better arrange this via email than discussing here so contact me

**Comment by dsnz** [comment]
Hi,

Working via teamviewer on your laptop is not very productive because of development setup, new programs I may need to install, information I will need about network, connection(s) etc.

So, let's try again.

Please follow the steps below.

1. restart your RAC
(I guess you can do this on the weekend or you do this periodically within the week;
if you can't do it then flush the shared pool in ALL the instances, you can do this by connecting directly to each instance by using a proper TNS entry)

2. connect to an instance directly with sqlplus (again by using a proper TNS entry)
and as user WEB_JSON

3. enter

```
alter session set tracefile_identifier = 'pljson';
alter session set events '10046 trace name context forever, level 12';
select * from table(pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
pljson_varray('ID', 'Ad', 'Tarih'),
table_mode => 'nested'
));
```

exactly as given and then logoff

4. find the latest trace files with 'pljson' in their name and in the trace directory of the instance you connected,
make a zip and post here

**Comment by serkangh** [comment]
Hi,
1.Now I flush the shared pool in ALL instances.
2.The second alter session command gives ORA-01031 Permission error. So I logged with another DBA account. And run the 3.step sql with this user. Because of alter session error.
3.I added schema name in front of the packages and varray names;
select * from table(web_json.pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
web_json.pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
web_json.pljson_varray('ID', 'Ad', 'Tarih'),
table_mode => 'nested'
));

The result trace files are as;
[Archive2.zip](https://github.com/pljson/pljson/files/2039476/Archive2.zip)

I can try to reboot RAC server this weekend.
May be this flush privileges on ALL instances can give you a trace idea.
I send you the same new trace file after reboot this weekend.

Thanks,
Serkan

**Comment by dsnz** [comment]
hmm, running from other account is not what I had in mind
since there maybe now issues with grants and synonyms (thought I give low probability to this)

please give permission for alter session (you can revoke it later) and run the statements within the WEB_JSON account

also I would like the init.ora (create it from spfile) for development databases and all production instances
if you feel they contain sensitive info then send directly by email

**Comment by dsnz** [comment]
@serkangh 
I'm sure you use 'cursor_sharing=force' in production database
please verify

**Comment by serkangh** [comment]
Yes you are right.
Development db setting is "EXACT"
Production db setting is "FORCE"

When I changed session setting to EXACT the error gone but a new error comes.
ORA-20110: JSON Path parse error: expected a string or an positive integer at 3

You committed a new push to GitHub 5 hours ago.
Before doing this tests. I already dropped user web_json and recreated schema and installed from your last commit!

What must I do now?
Thanks,
Serkan

**Comment by serkangh** [comment]
<img width="868" alt="screen shot 2018-05-28 at 00 27 55" src="https://user-images.githubusercontent.com/20636968/40590787-00ee1ab0-620e-11e8-83b4-af7a790d2d72.png">

**Comment by dsnz** [comment]
1. Your new error is because you run

```
select * from table(pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
pljson_varray('[].ID', '[].Ad', '[*].Tarih'),
pljson_varray('ID', 'Ad', 'Tarih'),
table_mode => 'nested'
));
```

instead of the correct

```
select * from table(pljson_table.json_table(
'[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]',
pljson_varray('[*].ID', '[*].Ad', '[*].Tarih'),
pljson_varray('ID', 'Ad', 'Tarih'),
table_mode => 'nested'
));
```
I had repeated the correct code before but seems you did not notice the difference
the wrong code is because you copied code from your post here but
when you posted you did not put this code inside 3 backquotes (`) and so it was mangled
(consult github markdown documentation)

2. my last release is a minor enhancement and will not solve your case entirely (see 3.)

3. for json_table() we use Oracle Data Cartridge interface
and the way this works now in Oracle has a side effect that it cannot be used with bind variables for all parameters
at least one of 'column_paths' or 'column_names' parameters must be literal (before the last release both had to be literals)

so the following works

```
declare
  bind1 pljson_varray := pljson_varray('[*].ID', '[*].Ad', '[*].Tarih');
  bind2 pljson_varray := pljson_varray('ID', 'Ad', 'Tarih');
  bind3 varchar2(100) := 'nested';
  row_count number;
begin
execute immediate 'select count(*)
from table(pljson_table.json_table(
''[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]'',
:bind1,
pljson_varray(''ID'', ''Ad'', ''Tarih''),
table_mode => :bind3
))'  into row_count using bind1, bind3;
  dbms_output.put_line(row_count);
end;
```

but if the column_names where changed to use bind variable 'bind2' then it would not work

I repeat that this is not a limitation of PLJSON but rather a result of how Oracle Data Cartridge works currently

of course running with cursor_sharing=force is impossible

so to summarize

a. run with cursor_sharing=EXACT either at the system level (recommended unless you have bad behaving applications, ie not using bind variables and making heavy use of memory, from a third party) or at the session level
b. use sql statements with literals
c.  or use sql statements with bind variables but one of 'column_paths' or 'column_names' parameters must be literal

**Comment by serkangh** [comment]
Thanks for your great help.
alter session with cursor_sharing commands in plsql gives pls-00103 errors. So solution (a) is not applicable.
But;
I finally solved with solution (c); bind variables and bulk collect to arrays is OK.

Thanks 
Serkan

**Comment by serkangh** [comment]
An interesting problem occurs when I convert plsql code to a procedure.
It started to give error ORA-00904.

a) Create a dummy type table
create table TEST123
(
  sinav_id     NUMBER not null,
  ad           VARCHAR2(1000) not null,
  tarih_txt    VARCHAR2(20) not null
);

b) This plsql can run gives the correct outputs.
declare
  bind1 web_json.pljson_varray := web_json.pljson_varray('[*].ID', '[*].Ad', '[*].Tarih');
  bind2 web_json.pljson_varray := web_json.pljson_varray('ID', 'Ad', 'Tarih');
  bind3 varchar2(100) := 'nested';
  row_count number;
  ID number;
  
   TYPE employee_ntt IS TABLE OF TEST123%ROWTYPE; 
   l_employees   employee_ntt; 

begin
execute immediate 'select ID,Ad,Tarih
from table(web_json.pljson_table.json_table(
''[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]'',
:bind1,
web_json.pljson_varray(''ID'', ''Ad'', ''Tarih''),
table_mode => :bind3
))' BULK COLLECT into l_employees using bind1, bind3;

  DBMS_OUTPUT.put_line (l_employees.COUNT); 
  DBMS_OUTPUT.put_line (l_employees(2).SINAV_ID); 
  DBMS_OUTPUT.put_line (l_employees(2).AD); 
  DBMS_OUTPUT.put_line (l_employees(2).TARIH_TXT); 
end;

c) But The below procedure is created and run from plsql it gives ORA-00904
CREATE OR REPLACE PROCEDURE PROC_SIL IS
  bind1 web_json.pljson_varray := web_json.pljson_varray('[*].ID', '[*].Ad', '[*].Tarih');
  bind2 web_json.pljson_varray := web_json.pljson_varray('ID', 'Ad', 'Tarih');
  bind3 varchar2(100) := 'nested';
  row_count number;
  ID number;
  
   TYPE employee_ntt IS TABLE OF TEST123%ROWTYPE; 
   l_employees   employee_ntt; 

begin
execute immediate 'select ID,Ad,Tarih
from table(web_json.pljson_table.json_table(
''[
{
"ID": 780,
"Ad": "1.Exam",
"Tarih": "05.032017"
},
{
"ID": 805,
"Ad": "2.Exam",
"Tarih": "09.072017"
}
]'',
:bind1,
web_json.pljson_varray(''ID'', ''Ad'', ''Tarih''),
table_mode => :bind3
))' BULK COLLECT into l_employees using bind1, bind3;

  DBMS_OUTPUT.put_line (l_employees.COUNT); 
  DBMS_OUTPUT.put_line (l_employees(2).SINAV_ID); 
  DBMS_OUTPUT.put_line (l_employees(2).AD); 
  DBMS_OUTPUT.put_line (l_employees(2).TARIH_TXT); 
end;

d) The error runner.
begin
  -- Call the procedure
  proc_sil;
end;

So I can not convert json to a table from a procedure. 
It is interesting.

**Comment by serkangh** [comment]
grantsandsynonyms.sql under src directory solves my problem.
Only I granted create public synonym to pljson owner schema.
Thanks for your great help.

**Comment by slingshot322** [comment]
@serkangh @dsnz I'm currently on 11.2.04 and installed latest PL/JSON version (2.3.0 as of this comment) and also had to use the `ALTER SESSION SET "CURSOR_SHARING" = EXACT;` command in order to resolve the ODCI error originally referenced in this issue.

Thanks to you both!

**Comment by dsnz** [comment]
good we solved this mystery 😄

**Comment by AslnurCakmak** [comment]
> grantsandsynonyms.sql under src directory solves my problem.
> Only I granted create public synonym to pljson owner schema.
> Thanks for your great help.

Hi  @serkangh ,



I'm getting the error too. When I tried the solution, I could not find a solution again.

Have any requests from the DBA team?

Or how did you go about solving this Problem?

Thanks, Regards.

**Comment by dsnz** [comment]
if you work in other schema then of course nothing works and you get this
error too among many others
you must make grants and synonyms (and directly , not thru roles)

you didn't specify your exact situation when you messaged me initially
that's why it's best to open a new issue on GitHub (after reading all
existing issues first)

if you work in the installation schema then the solution is in the
cursor_sharing parameter (issue #120) as I told you before

if you still have error then open a new issue detailing exact situation of
installation and error !
(without proper details we can't solve and also waste time)

**Comment by dsnz** [comment]
Please don't pollute old issues and email directly.
Try carefully all proposals as they are appropriate to your situation and if there is still problem then 
open a new issue.
